src/test/java/com/google/sps/servlets/RentPostsServletTest.java

```diff
@@ -0,0 +1,160 @@
+    String roomType = "SINGLE";
+    String startDate = "2020-07-10";
+    String title = "Test title";
+    setupDatabaseMocks();
+    setRequestParameters(description, endDate, leaseType, numRooms,
+      price, roomType, startDate, title);
+    Map<String, Object> expectedData = getExpectedRentData(description, endDate,
+      leaseType, numRooms, price, roomType, startDate, title);
+
+    rentPostsServlet.doPost(request, response);
+
+    verify(dbMock, Mockito.times(1)).collection(RENT_COLLECTION_NAME);
+    verify(collectionMock, Mockito.times(1)).add(expectedData);
+  }
+
+  @Test
+  public void testPost_requestHasUnparsableDates_postsWithNullDates() throws Exception {
+    String description = "Test description";
+    String endDate = "202020/20/10";
+    String leaseType = "YEAR_LONG";
+    String numRooms = "1";
+    String price = "100";
+    String roomType = "SINGLE";
+    String startDate = "07/10/2020";
+    String title = "Test title";
+    setupDatabaseMocks();
+    setRequestParameters(description, endDate, leaseType, numRooms,
+      price, roomType, startDate, title);
+    Map<String, Object> expectedData = 
+      getExpectedRentData(description, endDate, leaseType, numRooms, price, 
+        roomType, startDate, title);
+    expectedData.put(REQUEST_START_DATE, null);
+    expectedData.put(REQUEST_END_DATE, null);
+    
+    rentPostsServlet.doPost(request, response);
+
+    verify(dbMock, Mockito.times(1)).collection(RENT_COLLECTION_NAME);
+    verify(collectionMock, Mockito.times(1)).add(expectedData);
+  }
+  
+  private void setupDatabaseMocks() {
+    when(dbMock.collection(RENT_COLLECTION_NAME)).thenReturn(collectionMock);
+  }
+
+  /**
+  * Sets mock HTTP request's parameters to corresponding input values.
+  */
+  private void setRequestParameters(String description, String endDate, String leaseType,
+      String numRooms, String price, String roomType, String startDate, String title) {
+    when(request.getParameter(REQUEST_DESCRIPTION)).thenReturn(description);
+    when(request.getParameter(REQUEST_END_DATE)).thenReturn(endDate);
+    when(request.getParameter(REQUEST_LEASE_TYPE)).thenReturn(leaseType);
+    when(request.getParameter(REQUEST_NUM_ROOMS)).thenReturn(numRooms);
+    when(request.getParameter(REQUEST_PRICE)).thenReturn(price);
+    when(request.getParameter(REQUEST_ROOM_TYPE)).thenReturn(roomType);
+    when(request.getParameter(REQUEST_START_DATE)).thenReturn(startDate);
+    when(request.getParameter(REQUEST_TITLE)).thenReturn(title);
+  }
+
+
+  /**
+  * Sets mock HTTP request's parameters to corresponding input values.
+  * @return map of <database key, value> of expected data sent to database on a
+  *          servlet call to post()
+  */
+  private Map<String, Object> getExpectedRentData(String description, String endDate, String leaseType,
+      String numRooms, String price, String roomType, String startDate, String title) {
+    RentPost post = RentPost.builder()
```

Comment:
You can create RentPost object at top of the class and just change field you want to test using toBuilder() in test. This will save setting description, title and other properties everytime within each test.

Reply:
Moved builder creation and setting of parameters to the top of the class and calling .build() in each test.